### PR TITLE
manage.py: Add a --filter option to the CLI

### PR DESCRIPTION
**Pull request by Bjwebb**
A period separated argument is converted into the a '->' separated clause needed for postgres' JSON.

https://github.com/open-contracting/kingfisher-summarize/issues/190

To do:
- [x] Add more tests that filtering happens as expected

### CODE REVIEW

**Comment by Bjwebb** [comment]
@jpmckinney What do you think of the simple string replace used here for the where clause? Is this sufficient for now, or should we have more serious templating (e.g. jinja) here?

File: manage.py [hunk 1]
```diff
@@ -205,6 +206,20 @@ def validate_schema(ctx, param, value):
...
+def construct_extra_where_clause(cursor, filter_field, filter_value):
+    """
+    Returns part of a sql where clause, for given filter parameters.
+
+    :param str cursor: a pschopg2 database cursor
+    :param str filter_field: a period seperated field name, e.g. "tender.procurementMethod"
+    :param str filter_value: the value of the specified field, e.g. "direct"
+    """
+    path = [f"'{x}'" for x in filter_field.split('.')]
+    sql = 'AND d.data->' + '%s->'*(len(path)-1) + '>%s = %s'
```
Comment (jpmckinney) [inline manage.py:218]: This seems easier to read:

```suggestion
    sql = 'AND d.data' + '->%s' * (len(path) - 1) + '->>%s = %s'
```

File: manage.py [hunk 2]
```diff
@@ -205,6 +206,20 @@ def validate_schema(ctx, param, value):
...
 
+def construct_extra_where_clause(cursor, filter_field, filter_value):
+    """
+    Returns part of a sql where clause, for given filter parameters.
+
+    :param str cursor: a pschopg2 database cursor
+    :param str filter_field: a period seperated field name, e.g. "tender.procurementMethod"
+    :param str filter_value: the value of the specified field, e.g. "direct"
+    """
+    path = [f"'{x}'" for x in filter_field.split('.')]
```
Comment (jpmckinney) [inline manage.py:217]: I'm not sure I follow. `path` is only used for calling `len`? It seems we can just do `parts = filter_field.split('.')` and then reuse `parts` in `len(parts)` and in `mogrify`.

File: manage.py [hunk 3]
```diff
@@ -205,6 +206,20 @@ def validate_schema(ctx, param, value):
...
+    sql = 'AND d.data->' + '%s->'*(len(path)-1) + '>%s = %s'
+    escaped_sql = cursor.mogrify(sql, filter_field.split('.') + [filter_value])
```
Comment (jpmckinney) [inline manage.py:219]: The variables `sql` and `escaped_sql` would make more sense as `format_string` and `where_fragment`.

File: manage.py [hunk 4]
```diff
@@ -251,14 +268,19 @@ def add(ctx, collections, note, name, tables_only, field_counts_option, field_li
     NOTE is your name and a description of your purpose
     """
     logger = logging.getLogger('ocdskingfisher.summarize.add')
-    logger.info('Arguments: collections=%s note=%s name=%s tables_only=%s',
-                collections, note, name, tables_only)
+    logger.info('Arguments: collections=%s note=%s name=%s tables_only=%s filter_tuple=%s',
```
Comment (jpmckinney) [inline manage.py:271]: filter_tuple is the internal name, but I think `filter` will make more sense to users as it's the CLI option. That said, maybe we can name it `--where` (`where`).

File: manage.py [hunk 5]
```diff
@@ -85,6 +85,7 @@ def sql_files(directory, tables_only=False):
             content = f.read()
         if tables_only:
             content = re.sub(r'^(CREATE|DROP) VIEW', r'\1 TABLE', content, flags=flags)
+        content = content.replace('EXTRAWHERECLAUSE', extra_where_clause)
```
Comment (jpmckinney) [inline manage.py:88]: Once this is made into a comment as suggested, I think we can default `extra_where_clause`/`where_fragment` to `None` and then do this replacement `if where_fragment:`

File: manage.py [hunk 6]
```diff
@@ -251,14 +268,19 @@ def add(ctx, collections, note, name, tables_only, field_counts_option, field_li
...
-                collections, note, name, tables_only)
+    logger.info('Arguments: collections=%s note=%s name=%s tables_only=%s filter_tuple=%s',
+                collections, note, name, tables_only, filter_tuple)
 
     if not name:
         if len(collections) > 5:
             raise click.UsageError('--name is required for more than 5 collections')
         name = f"collection_{'_'.join(str(_id) for _id in sorted(collections))}"
 
+    if filter_tuple:
```
Comment (jpmckinney) [inline manage.py:279]: I'm not sure whether this logic belongs here or in `sql_files` (in which case we would just pass the `filter_tuple` / `where` argument around).

**Comment by jpmckinney** [comment]
> @jpmckinney What do you think of the simple string replace used here for the where clause? Is this sufficient for now, or should we have more serious templating (e.g. jinja) here?

I think it's fine - especially if my suggestion to change it to `-- WHERE_FRAGMENT` is applied.

File: manage.py (hunk 6 above)
Comment (Bjwebb) [reply manage.py:280]: If we're adding an extra option in future for SQL/JSON Path Language, then I think we want to keep this where it is, so we can convert both options to a where fragment to pass around.

File: manage.py (hunk 4 above)
Comment (Bjwebb) [reply manage.py]: I've stuck with `filter` for now, as I think that might make more sense alongside a SQL/JSON PathLanguage option.

**Review by jpmckinney — COMMENTED** [review]
pre-commit has been enabled organization-wide, but it is not configured for this repo. It's not a required check, so you don't need to worry about it.

File: manage.py [hunk 7]
```diff
@@ -241,8 +257,10 @@ def cli(ctx, quiet):
               help="Whether to add a field_list column to all summary tables (default false).")
 @click.option('--skip', multiple=True,
               help="Any SQL files to skip. Dependent files and final files will be skipped.")
+@click.option('--filter', "filter_tuple", nargs=2,
+              help="")
```
[the remark this answers is not in the document]
Comment (Bjwebb) [reply manage.py]: Yes, I've added multiple field support.

**Review by jpmckinney — APPROVED** [review]
I committed a nitpick (constructing a string with `+=` is slow).

**Pull request merged**